### DISCUSSION TITLE
[iOS] Missing axslc folder in the output package

### DISCUSSION THREAD

**Discussion by rh101** [axmolengine/axmol#1371] [General]
@halx99 I'm having a strange problem when building my app for iOS.  I've updated everything to work with Axmol 2.0, and that includes the `CMakeLists.txt`, but for some reason the axslc folder is missing from the output package.  

I can see that it is compiling the shaders during build, and that the generated files exist in the `build_ios_arm64/runtime` folder (including custom shaders), but for some reason it just doesn't copy them to the final iOS package.

If you can point me in the right direction as to where this copy actually occurs, then I may be able to track down the cause.  Generating a new project via `axmol new` works just fine for iOS, so there is something I'm missing when I updated my existing project from from Axmol 1.0 to 2.0.

I'm using this to generate the Xcode project:
`cmake -S . -B build_ios_arm64 -GXcode -DCMAKE_TOOLCHAIN_FILE=$AX_ROOT/1k/ios.cmake -DARCHS=arm64 -DENABLE_BITCODE=NO`

Am I missing something in that command?

EDIT: I'm going to try `axmol build -p ios -a arm64 -c` to see if there is any difference in the generated Xcode project.

**Comment by halx99** [comment]
EDIT: I'm going to try axmol build -p ios -a arm64 -c to see if there is any difference in the generated Xcode project.

This should be work, use nightly build cmake edition

**Reply by halx99** [reply]
require this feature: https://gitlab.kitware.com/halx99/cmake/-/tree/xcode-embed-resources/Tests/RunCMake/XcodeProject-Embed

**Reply by rh101** [reply]
`axmol build -p ios -a arm64 -c` worked, and the axslc exists in the output.

I'm assuming it's using the cmake version that axmol downloaded during setup instead of the installed version. Is that the case?

**Comment by halx99** [comment]
yes

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by rh101** [reply]
I understand. All good, thank you for clarifying.

**Comment by qoozta** [comment]
Can these files prebuilt and distribute to existing project and add to Resources folder manually?

**Reply by rh101** [reply]
Which version of Axmol is being used in the existing project?

**Reply by qoozta** [reply]
Axmol 1.0 without using CMake to compile (only used for first time Xcode project generation).
Just trying to upgrade to 2.0, also facing similar problem on Android which using asset pack.

**Reply by rh101** [reply]
If you don't mind, please explain why you don't want to use CMake after the initial project generation.  The reason I ask is if there is something you need to edit in the Xcode project manually after it's created, then perhaps there is a better way to do this via CMake that you may not be aware of.

Now, regarding the shaders, are you referring to all shaders, including the Axmol shaders, or just your own custom shaders?

If it's for your own custom shaders, then you could try adding them all to a new project, compile that app for each platform you require, and then just copy the shaders over to whatever path you need in your game.  Your own shaders can technically exist anywhere in your resources folder, since you always pass a relative path when you want to load them.  The engine shaders could also, technically, be anywhere in your resources folder, as long as you add the path to them to the search path in FileUtils.

**Reply by qoozta** [reply]
I just don't understand the good of using CMake on iOS, if I use Xcode directly, I just press run, then it compile and run the app on either physical device or simulator, if I have to use CMake to compile, do I need to use command to compile? How to debug via Xcode after compilation? I have been a Cocos user before the 2DX, so I prefer Xcode compilation.

**Reply by rh101** [reply]
> if I have to use CMake to compile, do I need to use command to compile?

I am a little confused by that statement.  You don't "compile" with CMake.  CMake is just used to generate the Xcode project, after which you load the generated project in Xcode, and that is where you compile/build the project.  The only time you need to re-run CMake is if you add new files to your project, which means CMake needs to regenerate your Xcode project files.

Can you please tell me why you think CMake compiles the project?  Is there something in the Axmol documentation that gives you the impression that CMake is doing the compilation?

**Reply by qoozta** [reply]
There has a command cocos compile as I remember?
The project I’m working need to update often and the SDKs may need to update often too, if each time adding files need to regenerate the project, it’s not good to me.

**Reply by rh101** [reply]
> There has a command cocos compile as I remember? 

Yes, and the equivalent for Axmol is `axmol build...`, but it's not CMake doing the compilation.  "axmol" is just a script that uses CMake to create the Xcode project, and then depending on what commands you pass to it, it can even trigger Xcode to compile your project.  If you pass the `-c` switch to it, then it simply creates the Xcode project files without doing the compile step.

> The project I’m working need to update often and the SDKs may need to update often too, if each time adding files need to regenerate the project, it’s not good to me.

~~What do you mean by "update often". What is it that you update?~~ Never mind, I misread.  That's just part of the normal development process.  You don't actually need to run CMake every time you add new files.  For example, you can add new source code files to your Xcode project directly, and continue development that way; that is not an issue.  The next time you re-run cmake, it'll just automatically add those files to your Xcode project too (assuming that you're using cmake `GLOB`/`GLOB_RECURSE`). If you don't use GLOB, then you'll need to add the source file references into the `CMakeLists.txt`.

As for the SDKs, *all* of them should be referenced from your `CMakeLists.txt`.  You should never be adding them to your Xcode project manually.

**Reply by qoozta** [reply]
That's the point, I have no reason to add files again in CMakeLists.txt, I've already added it in Xcode.
I will try to roll back to 1.0 then, it seems many works to do if migrate to 2.0.
